server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/MessageAppender.java

```diff
@@ -0,0 +1,113 @@
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.jmap.methods;
+
+import java.util.Date;
+import java.util.Optional;
+
+import javax.inject.Inject;
+import javax.mail.util.SharedByteArrayInputStream;
+
+import org.apache.james.jmap.model.Attachment;
+import org.apache.james.jmap.model.Keywords;
+import org.apache.james.jmap.model.MessageFactory;
+import org.apache.james.mailbox.AttachmentManager;
+import org.apache.james.mailbox.MailboxSession;
+import org.apache.james.mailbox.MessageManager;
+import org.apache.james.mailbox.exception.AttachmentNotFoundException;
+import org.apache.james.mailbox.exception.MailboxException;
+import org.apache.james.mailbox.model.AttachmentId;
+import org.apache.james.mailbox.model.Cid;
+import org.apache.james.mailbox.model.ComposedMessageId;
+import org.apache.james.mailbox.model.MessageAttachment;
+import org.apache.james.util.OptionalUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.fge.lambdas.Throwing;
+import com.github.fge.lambdas.functions.ThrowingFunction;
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ImmutableList;
+
+public class MessageAppender {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MessageAppender.class);
+
+    private final AttachmentManager attachmentManager;
+    private final MIMEMessageConverter mimeMessageConverter;
+
+    @Inject
+    public MessageAppender(AttachmentManager attachmentManager, MIMEMessageConverter mimeMessageConverter) {
+        this.attachmentManager = attachmentManager;
+        this.mimeMessageConverter = mimeMessageConverter;
+    }
+
+    public MessageFactory.MetaDataWithContent createMessageInMailbox(ValueWithId.MessageWithId.CreationMessageEntry createdEntry,
+                                                                     MessageManager mailbox,
```

Comment:
if it's a jmap API, why don't you take a MailboxId here ?

Reply:
Well SystemMailboxes provider directly gives MessageManager.

 -  No need to perform an extra lookup
 - We already have it in caller code.

Reply:
what caller code ? you don't know the caller, you are creating an API.
Or maybe it's a private class used by a single consumer ?
Otherwise, please make the API as consistent as you would like to find it as a developper browsing the API.

Reply:
benoit, if you want to keep this optimization you can provide the 2 methods, one with MessageId, and one with the MessageManager.

Reply:
Ok for duplicated methods of course.